### DISCUSSION TITLE
Best way to share data between /api routes and getStaticProps function?

### DISCUSSION THREAD

**Discussion by joaisa17** [vercel/next.js#39865] [Help] [answered]
I am making a site with content loaded from a CMS. I have made a caching system that clears appropriate data when it is updated, to ensure a minimal amount of requests to Sanity, as there is a monthly limit.

The cache itself works like intended, but the issue is that both getStaticProps and an API route I use to clear said cache, seem to run independently off each other, meaning that they use two different instances of the cache.

How do I get the API route and getStaticProps to share the same cache? Would I have to do an internal http request, and how would I do that?

The code for my cache class made in typescript is below, and is imported both in a page with getStaticProps, and an API route used for the client to fetch queries, such as image references. It's indended to be used in a context of Sanity documents.

```ts
import ms from 'ms';

type Data<T> = Sanity.WithId<T>|Sanity.WithId<T>[];

class ICache {
    private documentLinks = new Map<string, Set<string>>();
    private queryMap = new Map<string, {
        paths: string[];
        data: Data<any>;
        clearTimeout?: NodeJS.Timeout;
    }>();

    /**
     * Fetches the data for the given query. Returns undefined if no data was found.
     * Check the [cheat sheet](https://www.sanity.io/docs/query-cheat-sheet) for more info on how to use GROQ queries.
     * @param q GROQ query to fetch data for.
     * @returns 
     */
    get<T = any>(q: string, path: string): Data<T>|void {
        const cached = this.queryMap.get(q);
        if (cached) {
            const paths = [...cached.paths];
            !paths.includes(path) && paths.push(path);

            cached.clearTimeout && clearTimeout(cached.clearTimeout);
            this.queryMap.set(q, {
                paths,
                data: cached.data,
                clearTimeout: setTimeout(() => this.queryMap.delete(q), ms(process.env.CACHE_LIFETIME as string))
            });

            return cached.data as Data<T>;
        }
    }

    /**
     * Caches the given data to the given query.
     * @param q The query to cache.
     * @param path The path associated with the query.
     * @param data The data to set in the cache.
     */
    set<T = any>(q: string, path: string, data: Data<T>) {
        
        // Link each document if the data is an array.
        if (data instanceof Array) {
            data.forEach(d => {
                const links = this.documentLinks.get(d._id) ?? new Set();
                links.add(q);

                this.documentLinks.set(d._id, links);
            });
        }

        // Else, link the document to the query.
        else {
            const links = this.documentLinks.get(data._id) ?? new Set();
            links.add(q);


            this.documentLinks.set(data._id, links);
        }

        // Finally, cache the query.
        this.queryMap.set(q, {
            paths: [path],
            data
        });
    }

    /**
      Clears all queries that reference the given document ID.
      @param id ID of the document.
      @return all paths to be updated
    */
    clearDocumentID(id: string) {
        const paths = new Set<string>();

        // Loop through all queries that reference the document ID.
        this.documentLinks.get(id)?.forEach(q => {
            const cached = this.queryMap.get(q);

            if (cached) {
                // Insert the paths to be updated
                cached.paths.forEach(p => paths.add(p));

                // Delete the query from the cache, as it is no longer needed.
                this.queryMap.delete(q);
            }
        });

        // Finally, delete all links to the document ID.
        this.documentLinks.delete(id);

        return paths;
    }

    /**
     * Clears all queries associated with the given path.
     * @param path Path to clear queries for.
     * @return queries that were cleared.
     */
    clearPath(path: string): Set<string> {
        const queries = new Set<string>();

        this.queryMap.forEach((value, q) => {
            if (value.paths.includes(path)) {
                this.queryMap.delete(q);
                queries.add(q);
            }
        });

        return queries;
    }

    /** Clears the entire cache. */
    clearAll() {
        this.documentLinks = new Map();
        this.queryMap = new Map();
    }
}

const Cache = new ICache();
export default Cache;
```

**Comment by icyJoseph** [comment, marked as answer]
I think here the one issue is that `getStaticProps`, ignoring ISR, runs at build time, while api/routes, are executed at runtime.

In that scenario they cannot share any caching implemented by your code. You'd need a 3rd party, of your own perhaps, that handles the requests. For example an express server that responds the same thing.

> You could output JSON files I guess, or some other kind of static copy of the data, and access that via file system, but it is rather problematic, and might not even work in some hosts

At runtime when ISR is running, then the issue is that the pages are built with a worker, and nothing is really "live" on api/routes, because these are "serverless" by design. Again, a 3rd party server could handle serving the same data to both.

As requests arrive to api/routes, these are spun up, and depending on your hosting, they might be killed after a certain time. No JavaScript context is shared with the main Next.js app, and you shouldn't assume that it is.

> This depends a lot on how the Next.js app is hosted. 

This is a common pitfall in Next.js, for example if you have several pages building with data from Sanity, eventually, with enough pages, you might hit your rate-limits, and won't be able to build the app. Because each pages is built with their own worker, and they can't share between each other.

Of course, if you have only one instance of the server running, incoming requests, could share objects, but this is error prone. Your strategy of caching is a good attempt though, but even at runtime, keeping server side objects shared between different incoming client requests, can lead to memory leaks, race conditions and what not... but it is possible, under certain conditions. However, as said above, `getStaticProps` and the api/routes, as far as I know, can't really share JavaScript Contexts. I once tried to use workers and message passing, but it got really complex, and was not going anywhere. 

Often people implement a middle layer using a db such as Redis, which keeps sync with the CMS, and serves the front end app, both at build and runtime. This is also slightly better because it moves the responsibility of managing how much you connect with your CMS, as well as, eviction, revalidation, etc, away from the frontend app, which should just take care of rendering pages. Of course, it means yet additional costs...

**Reply by joaisa17** [reply]
Thank you for the detailed explanation!

Would I still be able to use the caching method in just the API if I'm planning to host this on Vercel? That is completely ignoring the cache when running getStaticProps, as it already only fetches at build time. It would at least let me share the data between clients.

This wouldn't be *as* optimised, but it would still be better than using no cache at all.
